### PR TITLE
Implementing the HCLU adversarial examples on GP for ART

### PR DESCRIPTION
**Pull request by KaGrosse**
# Description

Implementing the HCLU adversarial examples on GP for ART. This requires GPy for the gaussian processes. Contains unit tests for the new GPy class, the HCLU attack, and a corresponding notebook. 

Also adds a unit test for the decision tree attack. 

Fixes # 116

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [X ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- tests/attacks/test_HCLU.py
Getting iris down to binary setting for GP, testing whether attacks succeds for a fraction (see notebook for why it fails often), check if uncertainty for low uncertainty is indeed lower
- tests/attacks/test_decision_tree_attack.py
Check if targeted and untargeted both are gotten right
- tests/classifiers/test_GPy.py
Checks prediction, predictive uncertainty and gradients.

**Test Configuration**:
Mojave 10.14.5
Python 3.7.3
commit 
GPy: 1.9.8
scikit-learn: 0.21.0
scipy: 1.2.1

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/HCLU.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import numpy as np
+from scipy.optimize import minimize
+from art.attacks.attack import Attack
+from art.classifiers.GPy import GPyGaussianProcessClassifier
+logger = logging.getLogger(__name__)
+
+
+class HCLU(Attack):
+    """
+    # Implementation of the HCLU formulation in https://arxiv.org/abs/1812.02606 by Grosse et al.
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Format class docstring similar to other attacks in ART:
```suggestion
    Implementation of the High-Confidence-Low-Uncertainty (HCLU) adversarial example formulation by Grosse et al. (2018)
    Paper link:
        https://arxiv.org/abs/1812.02606 
```

File: art/attacks/HCLU.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import abc
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Please remove unused import `abc`.

File: art/attacks/HCLU.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import sys
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Please remove unused import `sys`.

File: art/attacks/HCLU.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+from art.attacks.attack import Attack
+from art.classifiers.GPy import GPyGaussianProcessClassifier
+logger = logging.getLogger(__name__)
+
+
+class HCLU(Attack):
+    """
+    # Implementation of the HCLU formulation in https://arxiv.org/abs/1812.02606 by Grosse et al.  
+    """
+    attack_params = ['conf', 'unc_increase']
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Should `minVal` and `maxVal` also be part of the list `attack_params`?

File: art/attacks/HCLU.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    def __init__(self, classifier, conf=0.95, unc_increase=100.0, minVal=0.0, maxVal=1.0):
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Could you please use lower case in variable names.
```suggestion
    def __init__(self, classifier, conf=0.95, unc_increase=100.0, min_val=0.0, max_val=1.0):
```

File: art/attacks/HCLU.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        :param conf: Confidence that examples should have, if there were to be classified as 1.0 maximally
+        :type conf: :float: 
+        :param unc_increase: Value uncertainty is allowed to deviate, where 1.0 is original value
+        :type unc_increase: :float:
+        :param minVal: minimal value any feature can take
+        :type minVal: :float:
+        :param maxVal: maximal value any feature can take
+        :type maxVal: :float:
+        """
+        self.classifier = classifier
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Please call first `__init__` of the super class: `super(HCLU, self).__init__(classifier=classifier)`
This replaces `self.classifier = classifier` which can be removed

File: art/attacks/HCLU.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                         'class_zero': class_zero, 'max_uncertainty': max_uncertainty}
+            cC = {'type': 'ineq', 'fun': constraint_conf, 'args': (init_args,)}
+            cU = {'type': 'ineq', 'fun': constraint_unc, 'args': (init_args,)}
+            args = {'args': init_args, 'orig': x[i].reshape(-1)}
+            # #finally, run optimization
+            adv_x[i] = minimize(minfun, adv_x[i], args=args,
+                                bounds=bounds, constraints=[cC, cU])['x']
+        return adv_x
+
+    def set_params(self, **kwargs):
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Could you please add `min_val` and `max_val` to `set_params`?

File: art/attacks/HCLU.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        :type unc_increase: :float:
+        :param minVal: minimal value any feature can take
+        :type minVal: :float:
+        :param maxVal: maximal value any feature can take
+        :type maxVal: :float:
+        """
+        self.classifier = classifier
+        if not isinstance(classifier, GPyGaussianProcessClassifier):
+            raise TypeError('Model must be a GPy Gaussian Process classifier!')
+        self.conf = conf
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Could you please replace lines 54-57 with code similar to:
```
params = {'conf': conf,
     'unc_increase': unc_increase,
     'min_val': min_val,
     'max_val': max_val
     }
self.set_params(**params)
```
That way the parameter check in `set_params` are applied also in `__init__`.

File: art/attacks/HCLU.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+class HCLU(Attack):
+    """
+    # Implementation of the HCLU formulation in https://arxiv.org/abs/1812.02606 by Grosse et al.  
+    """
+    attack_params = ['conf', 'unc_increase']
+
+    def __init__(self, classifier, conf=0.95, unc_increase=100.0, minVal=0.0, maxVal=1.0):
+        """
+        :param classifier: A trained model of type scikit decision tree.
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Please update the docstring this line.

File: art/attacks/HCLU.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        self.classifier = classifier
+        if not isinstance(classifier, GPyGaussianProcessClassifier):
+            raise TypeError('Model must be a GPy Gaussian Process classifier!')
+        self.conf = conf
+        self.unc_increase = unc_increase
+        self.minVal = minVal
+        self.maxVal = maxVal
+
+    def generate(self, x, y=None):
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]:
```suggestion
    def generate(self, x, y=None, **kwargs):
```

File: art/attacks/HCLU.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        :param kwargs: a dictionary of attack-specific parameters
+        :type kwargs: `dict`
+        :return: `True` when parsing was successful
+        """
+        super(HCLU, self).set_params(**kwargs)
+        if self.conf <= 0.5 or self.conf > 1.0:
+            raise ValueError("Confidence value has to be between 0.5 and 1.0.")
+        if self.unc_increase < 0.0:
+            raise ValueError(
+                "Uncertainty increase value has to be a positive nubmer.")
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Please fix typo.
```suggestion
                "Uncertainty increase value has to be a positive number.")
```

File: art/classifiers/GPy.py [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class GPyClassifier(Classifier):
+    """
+    Wrapper class for scikit-learn classifier models.
+    """
+
+    def __init__(self, model=None, channel_index=None, clip_values=None, defences=None, preprocessing=(0, 1)):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Please remove not required `channel_index` from `__init__` and docstring.

File: art/classifiers/GPy.py [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+import numpy as np
+
+from art.classifiers import Classifier, ClassifierGradients
+
+logger = logging.getLogger(__name__)
+
+
+class GPyClassifier(Classifier):
+    """
+    Wrapper class for scikit-learn classifier models.
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Update class docstring.

File: art/classifiers/GPy.py [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+    """
+
+    def __init__(self, model=None, channel_index=None, clip_values=None, defences=None, preprocessing=(0, 1)):
+        """
+        Create a `Classifier` instance from a GPy classifier model.
+
+        :param clip_values: Tuple of the form `(min, max)` representing the minimum and maximum values allowed
+               for features.
+        :type clip_values: `tuple`
+        :param model: scikit-learn classifier model.
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Update docstring. Is scikit-learn correct?

File: art/classifiers/GPy.py [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+               be divided by the second one.
+        :type preprocessing: `tuple`
+        """
+        super(GPyClassifier, self).__init__(clip_values=clip_values,
+                                            defences=defences, preprocessing=preprocessing)
+
+        self.model = model
+        self._input_shape = None
+
+    def fit(self, x, y, batch_size=128, nb_epochs=20, **kwargs):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Please remove `batch_size` and `nb_epochs`.

File: art/classifiers/GPy.py [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+        :param batch_size: Size of batches. Not used in this class
+        :type batch_size: `int`
+        :param nb_epochs: Number of iterations to be optimized.
+        :type nb_epochs: `int`
+        :type kwargs: `dict`
+        :return: `None`
+        """
+        raise NotImplementedError
+
+    def predict(self, x, logits=False, batch_size=128):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Please remove `batch_size` and `logits` from signature and docstring.

File: art/classifiers/GPy.py [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+        :param logits: `True` if the prediction should be done at the logits layer.
+        :type logits: `bool`
+        :param batch_size: Size of batches. Not used in this function.
+        :type batch_size: `int`
+        :return: Array of predictions of shape `(nb_inputs, self.nb_classes)`.
+        :rtype: `np.ndarray`
+        """
+        raise NotImplementedError
+
+    def predict_uncertainty(self, x, logits=False, batch_size=128):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Please remove `batch_size` from signature and docstring.

File: art/classifiers/GPy.py [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+                ind = self.predict(x[i].reshape(1, -1))[0, c]
+                sur = self.predict(np.repeat(x[i].reshape(1, -1),
+                                             np.shape(x)[1], 0) + eps*np.eye(np.shape(x)[1]))[:, c]
+                grads[i, c] = ((sur-ind)*eps).reshape(1, -1)
+        if not(label is None):
+            return grads[:, label, :].reshape(np.shape(x)[0], 1, np.shape(x)[1])
+        else:
+            return grads
+
+    def get_activations(self, x, layer, batch_size):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Remove this function because it's not required by the used classifier interfaces.

File: art/classifiers/GPy.py [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+        grads = np.zeros(np.shape(x))
+        for i in range(np.shape(x)[0]):
+            # 1.0 - to mimic loss, [0,np.argmax] to get right class
+            ind = 1.0-self.predict(x[i].reshape(1, -1))[0, np.argmax(y[i])]
+            sur = 1.0-self.predict(np.repeat(x[i].reshape(1, -1), np.shape(x)[1], 0)
+                                   + eps*np.eye(np.shape(x)[1]))[:, np.argmax(y[i])]
+            grads[i] = ((sur-ind)*eps).reshape(1, -1)
+        return grads
+
+    def predict(self, x, logits=False, batch_size=128):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Please remove `batch_size` from signature and docstring.

File: art/classifiers/GPy.py [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+        if logits:  # output the mon-squashed version
+            out[:, 0] = self.model.predict_noiseless(x)[0].reshape(-1)
+            out[:, 1] = -1.0*out[:, 0]
+            return out
+        # output normal prediction, scale up to two values
+        out[:, 0] = self.model.predict(x)[0].reshape(-1)
+        out[:, 1] = 1.0-out[:, 0]
+        return out
+
+    def predict_uncertainty(self, x, logits=False, batch_size=128):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Please remove batch_size from signature and docstring.

File: art/classifiers/GPy.py [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+        :type logits: `bool`
+        :param batch_size: Size of batches. Not used in this function.
+        :type batch_size: `int`
+        :return: Array of uncertainty predictions of shape `(nb_inputs, self.nb_classes)`.
+        :rtype: `np.ndarray`
+        """
+        # geturn uncertainty, only in one dimension
+        return self.model.predict_noiseless(x)[1]
+
+    @property
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Remove this function because it's not required by the used classifier interfaces.

File: art/classifiers/GPy.py [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+        :rtype: `list`
+
+        .. warning:: `layer_names` tries to infer the internal structure of the model.
+                     This feature comes with no guarantees on the correctness of the result.
+                     The intended order of the layers tries to match their order in the model, but this is not
+                     guaranteed either.
+        """
+        raise NotImplementedError
+
+    def set_learning_phase(self, train):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Remove this function because it's not required by the used classifier interfaces.

File: art/classifiers/GPy.py [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+                grads[i, c] = ((sur-ind)*eps).reshape(1, -1)
+        if not(label is None):
+            return grads[:, label, :].reshape(np.shape(x)[0], 1, np.shape(x)[1])
+        else:
+            return grads
+
+    def get_activations(self, x, layer, batch_size):
+        raise NotImplementedError
+
+    def loss_gradient(self, x, y):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Please match signature of `loss_gradient` in base class.
```suggestion
    def loss_gradient(self, x, y, **kwargs):
```

File: art/classifiers/GPy.py [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+        :type x: `np.ndarray`
+        :param logits: `True` if the prediction should be done without squashing function.
+        :type logits: `bool`
+        :param batch_size: Size of batches. Not used in this function.
+        :type batch_size: `int`
+        :return: Array of predictions of shape `(nb_inputs, self.nb_classes)`.
+        :rtype: `np.ndarray`
+        """
+        out = np.zeros((np.shape(x)[0], 2))
+        if logits:  # output the mon-squashed version
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]:
```suggestion
        if logits:  # output the non-squashed version
```

File: tests/classifiers/test_GPy.py [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+from sklearn.svm import SVC, LinearSVC
+
+from art.classifiers import GPyGaussianProcessClassifier
+from art.utils import load_dataset
+
+logger = logging.getLogger('testLogger')
+np.random.seed(seed=1234)
+
+
+class TestScikitlearnDecisionTreeClassifier(unittest.TestCase):
```
Comment (beat-buesser) [inline tests/classifiers/test_GPy.py]: Please update the name of the test class (GPy?).

File: art/classifiers/GPy.py [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+        :return: Array of uncertainty predictions of shape `(nb_inputs, self.nb_classes)`.
+        :rtype: `np.ndarray`
+        """
+        raise NotImplementedError
+
+    def save(self, filename, path=None):
+        self.model.save_model(filename, save_data=False)
+
+
+class GPyGaussianProcessClassifier(GPyClassifier, ClassifierGradients):
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Would it be possible to combine the two classes in this module into 1 class or are they split into 2 classes for a future extension to include `sklearn.gaussian_process.GaussianProcessClassifier`?

File: art/classifiers/GPy.py [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+    """
+    Wrapper class for scikit-learn Decision Tree Classifier models.
+    """
+
+    def __init__(self, model=None, channel_index=None, clip_values=None, defences=None, preprocessing=(0, 1)):
+        """
+        Create a `Classifier` instance from a scikit-learn Decision Tree Classifier model.
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Update docstring (scikit-learn, Decision Tree, etc.)

File: art/classifiers/GPy.py [hunk 28]
```diff
@@ -0,0 +1,254 @@
...
+        """
+        raise NotImplementedError
+
+    def save(self, filename, path=None):
+        self.model.save_model(filename, save_data=False)
+
+
+class GPyGaussianProcessClassifier(GPyClassifier, ClassifierGradients):
+    """
+    Wrapper class for scikit-learn Decision Tree Classifier models.
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]: Update docstring.

File: art/attacks/HCLU.py [hunk 29]
```diff
@@ -0,0 +1,118 @@
...
+            bounds.append((self.minVal, self.maxVal))
+        for i in range(np.shape(x)[0]):  # go though data amd craft
+            # get properties for attack
+            max_uncertainty = self.unc_increase * \
+                self.classifier.predict_uncertainty(adv_x[i].reshape(1, -1))
+            class_zero = not self.classifier.predict(
+                adv_x[i].reshape(1, -1))[0, 0] < 0.5
+            init_args = {'classifier': self.classifier,
+                         'class_zero': class_zero, 'max_uncertainty': max_uncertainty}
+            cC = {'type': 'ineq', 'fun': constraint_conf, 'args': (init_args,)}
```
Comment (beat-buesser) [inline art/attacks/HCLU.py]: Would it be possible to find a snake_case naming style variable name for `cC` and `cU`?

File: art/attacks/HCLU.py (hunk 9 above)
Comment (KaGrosse) [reply art/attacks/HCLU.py]: Hi Beat, thanks for you feedback - you mean, updating it with the default value, or with the new variable names?  Or something else?

Comment (beat-buesser) [reply art/attacks/HCLU.py]: I was thinking about `scikit decision tree`, is it still correct?

File: art/attacks/HCLU.py (hunk 7 above)
Comment (KaGrosse) [reply art/attacks/HCLU.py]: Added below check that max>min, should be solved.

File: art/attacks/HCLU.py (hunk 9 above)
Comment (KaGrosse) [reply art/attacks/HCLU.py]: Yep, fixed. Sort of overlooked that -.-

File: art/classifiers/GPy.py (hunk 12 above)
Comment (KaGrosse) [reply art/classifiers/GPy.py]: Is done, strangely not shown here O.o

File: art/classifiers/GPy.py (hunk 18 above)
Comment (KaGrosse) [reply art/classifiers/GPy.py]: Is removed.

File: art/classifiers/GPy.py (hunk 19 above)
Comment (KaGrosse) [reply art/classifiers/GPy.py]: is removed.

File: art/classifiers/GPy.py (hunk 20 above)
Comment (KaGrosse) [reply art/classifiers/GPy.py]: is removed.

**Pull request merged**